File: jvm/sargon-android/src/main/java/com/radixdlt/sargon/extensions/FactorSource.kt
package com.radixdlt.sargon.extensions

import com.radixdlt.sargon.ArculusCardFactorSource
import com.radixdlt.sargon.DeviceFactorSource
import com.radixdlt.sargon.FactorSource
import com.radixdlt.sargon.FactorSourceId
import com.radixdlt.sargon.FactorSourceKind
import com.radixdlt.sargon.HostInfo
import com.radixdlt.sargon.LedgerHardwareWalletFactorSource
import com.radixdlt.sargon.MnemonicWithPassphrase
import com.radixdlt.sargon.OffDeviceMnemonicFactorSource
import com.radixdlt.sargon.PasswordFactorSource
import com.radixdlt.sargon.SpotCheckInput
import com.radixdlt.sargon.factorSourceSupportsBabylon
import com.radixdlt.sargon.factorSourceSupportsOlympia
import com.radixdlt.sargon.factorSourceName
import com.radixdlt.sargon.factorSourcePerformSpotCheck
import com.radixdlt.sargon.newDeviceFactorSourceBabylon
import com.radixdlt.sargon.newDeviceFactorSourceOlympia

val FactorSource.id: FactorSourceId
    get() = when (this) {
        is FactorSource.Device -> value.id.asGeneral()
        is FactorSource.Ledger -> value.id.asGeneral()
        is FactorSource.ArculusCard -> value.id.asGeneral()
        is FactorSource.OffDeviceMnemonic -> value.id.asGeneral()
//        is FactorSource.SecurityQuestions -> value.id.asGeneral()
//        is FactorSource.TrustedContact -> value.id.asGeneral()
        is FactorSource.Password -> value.id.asGeneral()
    }

val FactorSource.kind: FactorSourceKind
    get() = when (this) {
        is FactorSource.Device -> value.kind
        is FactorSource.Ledger -> value.kind
        is FactorSource.ArculusCard -> value.kind
        is FactorSource.OffDeviceMnemonic -> value.kind
//        is FactorSource.SecurityQuestions -> value.kind
//        is FactorSource.TrustedContact -> value.kind
        is FactorSource.Password -> value.kind
    }

fun DeviceFactorSource.asGeneral() = FactorSource.Device(value = this)
fun LedgerHardwareWalletFactorSource.asGeneral() = FactorSource.Ledger(value = this)
fun ArculusCardFactorSource.asGeneral() = FactorSource.ArculusCard(value = this)
fun OffDeviceMnemonicFactorSource.asGeneral() = FactorSource.OffDeviceMnemonic(value = this)
//fun SecurityQuestionsNotProductionReadyFactorSource.asGeneral() =
//    FactorSource.SecurityQuestions(value = this)
//fun TrustedContactFactorSource.asGeneral() = FactorSource.TrustedContact(value = this)
fun PasswordFactorSource.asGeneral() = FactorSource.Password(value = this)

fun FactorSource.Device.Companion.olympia(
    mnemonicWithPassphrase: MnemonicWithPassphrase,
    hostInfo: HostInfo
) = newDeviceFactorSourceOlympia(
    mnemonicWithPassphrase = mnemonicWithPassphrase,
    hostInfo = hostInfo
).asGeneral()

fun FactorSource.Device.Companion.babylon(
    mnemonicWithPassphrase: MnemonicWithPassphrase,
    hostInfo: HostInfo
) = newDeviceFactorSourceBabylon(
    mnemonicWithPassphrase = mnemonicWithPassphrase,
    hostInfo = hostInfo
).asGeneral()

fun FactorSource.spotCheck(
    input: SpotCheckInput
) = factorSourcePerformSpotCheck(this, input)

val FactorSource.supportsOlympia: Boolean
    get() = factorSourceSupportsOlympia(factorSource = this)

val FactorSource.supportsBabylon: Boolean
    get() = factorSourceSupportsBabylon(factorSource = this)

val FactorSource.name: String
    get() = factorSourceName(factorSource = this)

val DeviceFactorSource.kind: FactorSourceKind
    get() = FactorSourceKind.DEVICE

val LedgerHardwareWalletFactorSource.kind: FactorSourceKind
    get() = FactorSourceKind.LEDGER_HQ_HARDWARE_WALLET

val ArculusCardFactorSource.kind: FactorSourceKind
    get() = FactorSourceKind.ARCULUS_CARD

val OffDeviceMnemonicFactorSource.kind: FactorSourceKind
    get() = FactorSourceKind.OFF_DEVICE_MNEMONIC

//val SecurityQuestionsNotProductionReadyFactorSource.kind: FactorSourceKind
//    get() = FactorSourceKind.SECURITY_QUESTIONS

//val TrustedContactFactorSource.kind: FactorSourceKind
//    get() = FactorSourceKind.TRUSTED_CONTACT

val PasswordFactorSource.kind: FactorSourceKind
    get() = FactorSourceKind.PASSWORD

File: crates/uniffi/uniffi_SPLIT_ME/src/profile/v100/factors/factor_source.rs
use crate::prelude::*;
use sargon::BaseBaseIsFactorSource;
use sargon::FactorSource as InternalFactorSource;

decl_vec_samples_for!(FactorSources, FactorSource);

#[derive(Clone, PartialEq, Eq, Hash, uniffi::Enum)]
pub enum FactorSource {
    Device {
        value: DeviceFactorSource,
    },

    Ledger {
        value: LedgerHardwareWalletFactorSource,
    },

    OffDeviceMnemonic {
        value: OffDeviceMnemonicFactorSource,
    },

    ArculusCard {
        value: ArculusCardFactorSource,
    },

    Password {
        value: PasswordFactorSource,
    },
}

delegate_debug_into!(FactorSource, InternalFactorSource);

impl FactorSource {
    pub fn into_internal(&self) -> InternalFactorSource {
        self.clone().into()
    }
}

impl From<InternalFactorSource> for FactorSource {
    fn from(value: InternalFactorSource) -> Self {
        match value {
            InternalFactorSource::Device { value } => FactorSource::Device {
                value: value.into(),
            },
            InternalFactorSource::Ledger { value } => FactorSource::Ledger {
                value: value.into(),
            },
            InternalFactorSource::OffDeviceMnemonic { value } => {
                FactorSource::OffDeviceMnemonic {
                    value: value.into(),
                }
            }
            InternalFactorSource::ArculusCard { value } => {
                FactorSource::ArculusCard {
                    value: value.into(),
                }
            }
            InternalFactorSource::Password { value } => {
                FactorSource::Password {
                    value: value.into(),
                }
            }
            InternalFactorSource::TrustedContact { value: _ } => {
                panic!("TrustedContact not yet supported in the Wallet")
            }
            InternalFactorSource::SecurityQuestions { value: _ } => {
                panic!("SecurityQuestions not yet supported in the Wallet")
            }
        }
    }
}

impl From<FactorSource> for InternalFactorSource {
    fn from(val: FactorSource) -> Self {
        match val {
            FactorSource::Device { value } => InternalFactorSource::Device {
                value: value.into_internal(),
            },
            FactorSource::Ledger { value } => InternalFactorSource::Ledger {
                value: value.into_internal(),
            },
            FactorSource::OffDeviceMnemonic { value } => {
                InternalFactorSource::OffDeviceMnemonic {
                    value: value.into_internal(),
                }
            }
            FactorSource::ArculusCard { value } => {
                InternalFactorSource::ArculusCard {
                    value: value.into_internal(),
                }
            }
            FactorSource::Password { value } => {
                InternalFactorSource::Password {
                    value: value.into_internal(),
                }
            }
        }
    }
}

#[uniffi::export]
pub fn factor_source_to_string(factor_source: &FactorSource) -> String {
    factor_source.into_internal().to_string()
}

#[uniffi::export]
pub fn factor_source_supports_olympia(factor_source: &FactorSource) -> bool {
    factor_source.into_internal().supports_olympia()
}

#[uniffi::export]
pub fn factor_source_supports_babylon(factor_source: &FactorSource) -> bool {
    factor_source.into_internal().supports_babylon()
}

#[uniffi::export]
pub fn factor_source_name(factor_source: &FactorSource) -> String {
    factor_source.into_internal().name()
}

#[uniffi::export]
pub fn new_factor_source_sample() -> FactorSource {
    InternalFactorSource::sample().into()
}

#[uniffi::export]
pub fn new_factor_source_sample_other() -> FactorSource {
    InternalFactorSource::sample_other().into()
}
